attribution/attribution_metrics.py

```diff
@@ -33,105 +33,30 @@ def token_prob_difference(
-
-    return max_difference
-
-
-def get_sentence_embeddings(
-    sentence: str, model: PreTrainedModel, tokenizer: PreTrainedTokenizer
-) -> Tuple[np.ndarray, np.ndarray]:
-    inputs = tokenizer(sentence, return_tensors="pt")
-    embeddings = model.transformer.wte(inputs["input_ids"])  # Get the embeddings
-    embeddings = embeddings.detach().numpy().squeeze(axis=0)
-    return embeddings.mean(axis=0), embeddings
-
-
 def cosine_similarity_attribution(
-    original_output_choice: openai.types.chat.chat_completion.Choice,
-    perturbed_output_choice: openai.types.chat.chat_completion.Choice,
-    model: PreTrainedModel,
+    original_output_str: str,
+    perturbed_output_str: str,
+    token_embeddings: np.ndarray,
     tokenizer: PreTrainedTokenizer,
 ) -> Tuple[float, np.ndarray]:
     # Extract embeddings
-    initial_output_sentence_emb, initial_output_token_embs = get_sentence_embeddings(
-        original_output_choice.message.content, model, tokenizer
-    )
-    perturbed_output_sentence_emb, perturbed_output_token_embs = get_sentence_embeddings(
-        perturbed_output_choice.message.content, model, tokenizer
-    )
-
-    # Reshape embeddings
-    initial_output_sentence_emb = initial_output_sentence_emb.reshape(1, -1)
-    perturbed_output_sentence_emb = perturbed_output_sentence_emb.reshape(1, -1)
 
-    # Calculate similarities
-    self_similarity = float(
-        cosine_similarity(initial_output_sentence_emb, initial_output_sentence_emb)
-    )
-    sentence_similarity = float(
-        cosine_similarity(initial_output_sentence_emb, perturbed_output_sentence_emb)
-    )
+    original_token_ix = tokenizer.encode(original_output_str, return_tensors="pt", add_special_tokens=False)
+    perturbed_token_ix = tokenizer.encode(perturbed_output_str, return_tensors="pt", add_special_tokens=False)
+    initial_tokens = [tokenizer.decode(t) for t in original_token_ix.squeeze(axis=0)]
 
-    # Calculate token similarities for shared length
-    shared_length = min(initial_output_token_embs.shape[0], perturbed_output_token_embs.shape[0])
-    token_similarities_shared = cosine_similarity(
-        initial_output_token_embs[:shared_length],
-        perturbed_output_token_embs[:shared_length],
-    ).diagonal()
-
-    # Pad token similarities to match initial token embeddings shape
-    token_similarities = np.pad(
-        token_similarities_shared,
-        (0, initial_output_token_embs.shape[0] - shared_length),
-    )
+    original_output_emb = token_embeddings[original_token_ix].reshape(-1, token_embeddings.shape[-1])
+    perturbed_output_emb = token_embeddings[perturbed_token_ix].reshape(-1, token_embeddings.shape[-1])
 
-    # Return difference in sentence similarity and token similarities
-    return self_similarity - sentence_similarity, 1 - token_similarities
+    cd = 1-cosine_similarity(original_output_emb, perturbed_output_emb)
+    token_distance = cd.min(axis=-1)    
```

Comment:
Sure! So, cosine_similarity actually gives us a pairwise score between every unperturbed output embedding and every perturbed output embedding. So if unperturbed output embeddings is shape (10,768), and the perturbed is (6, 768), cosine_distance = 1-cosine_similarity(original_output_emb, perturbed_output_emb) gives us a matrix of shape (10, 6). And we want the minimum values for each _unperturbed_ output embedding, so we take the min over the last dimension to give an attribution vector length 10 – i.e. one element for each token in the original unperturbed output.

Reply:
Yeah, we can't really capture repetitions - it's not recognised as a change in the output, because we pick the lowest distance for each embedding. I'm not sure how to get around this though. 
<img width="1919" alt="image" src="https://github.com/leap-laboratories/PIZZA/assets/6306764/658cda21-29ab-4142-b28b-fb9efbb417f6">

Reply:
Thanks! That makes sense. Yeah we're computing the pairwise similarities instead of comparing elements in the same position as before. I think that's good although I agree it has some limitations -- repeated tokens and frequent tokens which appear twice for different reasons may affect the attribution score. I think we can leave it as is for now 👍🏼